ledger/participant-state/kvutils/src/test/suite/scala/com/daml/ledger/participant/state/kvutils/committer/transaction/TransactionCommitterSpec.scala

```diff
@@ -595,58 +595,125 @@ class TransactionCommitterSpec extends AnyWordSpec with Matchers with MockitoSug
+      } else {
+        builder.add(lookup)
+      }
+      builder.buildSubmitted()
+    }
+
+    val mykey = {
+      val aCreateNode = createNode("#dummy")
+      Conversions.encodeContractKey(aCreateNode.templateId, aCreateNode.key.get.key)
+    }
 
-      val lookupByKeyNotFound =
-        newLookupByKeySubmittedTransaction(found = false)
-      val lookupByKeyFound1 = newLookupByKeySubmittedTransaction(found = true)
-      val lookupByKeyFound2 = newLookupByKeySubmittedTransaction(found = true)
+    "return Inconsistent when a contract key resolves to a different contract ID than submitted by a participant" in {
+
+      val cases =
+        Seq((false, Some(s"#$freshContractId")), (true, Some(s"#$freshContractId")), (true, None))
+          .flatMap { case (found, contractIdAtCommitter) =>
+            Seq(false, true).map(inRollback =>
+              newLookupByKeySubmittedTransaction(found, inRollback) -> contractIdAtCommitter
+            )
+          }
 
-      val cases = Table(
+      val casesTbl = Table(
         "transaction" -> "contractIdAtCommitter",
-        lookupByKeyNotFound -> Some(s"#$freshContractId"),
-        lookupByKeyFound1 -> Some(s"#$freshContractId"),
-        lookupByKeyFound2 -> None,
+        cases: _*
       )
 
-      val key = {
-        val aCreateNode = createNode("#dummy")
-        Conversions.encodeContractKey(aCreateNode.templateId, aCreateNode.key.get.key)
-      }
-      forAll(cases) { (transaction: SubmittedTransaction, contractIdAtCommitter: Option[String]) =>
-        val context = commitContextWithContractStateKeys(
-          key -> contractIdAtCommitter
-        )
-        val result =
-          ContractKeysValidation
-            .validateKeys(transactionCommitter)
-            .apply(
-              context,
-              DamlTransactionEntrySummary(createTransactionEntry(List("Alice"), transaction)),
-            )
-        result shouldBe a[StepStop]
-        val rejectionReason =
-          getTransactionRejectionReason(result).getInconsistent.getDetails
-        rejectionReason should startWith("InconsistentKeys")
+      forAll(casesTbl) {
+        (transaction: SubmittedTransaction, contractIdAtCommitter: Option[String]) =>
+          val context = commitContextWithContractStateKeys(
+            mykey -> contractIdAtCommitter
+          )
+          val result = validate(context, transaction)
+          result shouldBe a[StepStop]
+          val rejectionReason =
+            getTransactionRejectionReason(result).getInconsistent.getDetails
+          rejectionReason should startWith("InconsistentKeys")
       }
     }
 
+    "return DuplicateKeys when two local contracts conflict" in {
+      val builder = TransactionBuilder()
+      builder.add(createNode(s"#$freshContractId"))
+      builder.add(createNode(s"#$freshContractId"))
+      val transaction = builder.buildSubmitted()
+      val context = commitContextWithContractStateKeys(mykey -> None)
```

Comment:
Is this necessary?

Reply:
Depends™. If there is no entry, kvutils failed to track it’s inputs correctly. I think the test still fails but that’s not what we want to test here. This entry means kvutils knows it’s an input and the key does not exist before.